### ISSUE TITLE
new #send_file behavior is error-prone in globbed routes

### ISSUE THREAD

**Issue by tercenya** [hanami/controller#210]
version: 1.0.0-beta1 

https://github.com/hanami/controller/commit/4a1e1af2c6da5b6adf8d9d3c17b397dbd3487b5e restricts send_file to files only in public_directory, and allows the router to attempt other routes.  This causes an nasty problem that isn't obvious.

When using a glob route, say:
```
  get '/path/:id(.:format), to 'example#show'
```
and the `#show` action does a `#send_file` that isn't in the public folder, the return value is has a "X-Cascade"=>"pass" header.  This entails a second attempt on the same route, except :format isn't globbed out (:id is now "#{id}.#{format}").  This will almost certainly fail with a 404/406 on a well written controller, and the original "File Not Found" 404 error message is lost to the new "ID is invalid".  Looking for a previous failure after handing the controller a new set of (invalid) params doesn't seem like a practice I want in my controllers.

Moreover, a controller that authoritatively says "send this file as a body" shouldn't fail silently - it should be a hard 404 or 500 if unable to do so (in [some|many|most] cases), just like a missing template.   If people also need a soft-failing `#send_file`, perhaps the weaker version should be `#attempt_send_file`, or alternatively move the more firm version to `#send_file!`, so we don't have failures coming from other controllers or another pass through the same controller with bogus data.

**Comment by tercenya** [comment]
**tl;dr**: Using `#unsafe_send_file` from [216](https://github.com/hanami/controller/pull/216) will sidestep the issue for now, but this the reason I opened this issue separately was to discuss whether there should also be a `#send_file!` version that allows you to have the benefits of the whitelisted path checking, without the downside of unwanted route cascading.

long version:

I've made some branches that explain this issue better, but there is some room for discussion as to the nature of the problem.

First up, the [send_file_with_glob](https://github.com/hanami/controller/compare/unsafe_send_file...tercenya:send_file_with_glob) branch.  The changes of note are that I put the more specific '/files/unsafe/' route above the glob route<sup id="a1">[1](#f1)</sup>, and I added a "glob everything" route last that always returns 202.  Two tests now fail: [doesn't send file outside of public directory](https://github.com/tercenya/hanami-controller/blob/send_file_with_glob/test/integration/send_file_test.rb#L64) and [if file doesn't exist, respond 404](https://github.com/tercenya/hanami-controller/blob/send_file_with_glob/test/integration/send_file_test.rb#L71).  Both fail "silently" because our send_file's 404 response gets 'eaten' by the glob route.

If you wanted a more stringent `send_file`, (e.g. one that *actually* returns 404 when the file can't be sent), you can use this:
```ruby
      # Same as #send_file, but if we fail, we will stop routing and return a 404
      # @TODO: not very DRY with regard to #send_file.
      def send_file!(path)
        result = File.new(path, self.class.configuration.public_directory).call(@_env)
        headers.merge!(result[1])
        
        if result[0] == 404
          headers.delete('X-Cascade')
          halt 404
        else
          halt result[0], result[2]
        end
      end
```

One can argue this is the intended behavior of route matching, but this gets uglier (and more persuasive) if you implement the #show route closer to a real-world use case, instead of that greedy glob.

Now take a look at [send_file_with_formatting_globs](https://github.com/hanami/controller/compare/unsafe_send_file...tercenya:send_file_with_formatting_globs).  The #show route does more work, but the idea here is that a '/:id(.:format)' route is supposed to be looking for Resource[X] and respond in a potential format. The Expected behavior: 

- when ID is invalid, we should get a 400. 
- If :format is always invalid, we should get a 406.
-  if :format is unspecified, choose a default (I've picked :json because the code path here is clearer, but :html is also common).
- If Resource[ID] is valid, but we we just don't have it available in said :format, get a 404. We'll be using send_file, because our resources are just flat files (in my $JOB, these are photos).

I've added tests for these routes, and the this is fine up until bullet #4.  We get a valid resource, then attempt `send_file`: [get /files/501.json](https://github.com/tercenya/hanami-controller/blob/6ba287f6f5394e67909946d964b543751a789ce6/test/integration/send_file_test.rb#L124)  what's most curious is that we get a 400, not a 404.  This is because routing has caused our controller to run a *second time*, and now we're looking for an (invalid) ID of '500.json' with format: nil.   Wrong code path, we need to get that 404 out first!

When I upgraded from 0.9.X to 1.0.0-beta, `#send_file` went from "returning files with content (in an unsafe folder)" to silently failing with a 404, but that 404 isn't returned anywhere because glob routes eat that for lunch; instead all I see is 400/406/500, complaining about invalid IDs and unacceptable formats, which has nothing to do with the "missing files." 

You could say this is all _working as intended_ and that my controller should have checked that the target `.exists?` first AND is located in the whitelisted public folder, otherwise return 404.  Which is *exactly* the behavior of new `#send_file`, not withstanding the side-effects of using X-Cascade.  Hence the request for a `#send_file!` variant, and a note somewhere in the documentation that 'doing `#send_file` has changed in 1.0.0 to be better, but it's more challenging - here's what you need to consider'

<sup id="f1">1</sup>: although the change makes no difference here, there is a problem shortly. ideally you shouldn't be calling the #show route with { id: 'unsafe' }. If that controller does hard check on the validity of :id, specific routes must be ahead of non-specific routes.  [↩](#a1)
